File: py/binary-indexed-tree.py
class BIT:

    """ Binary Indexed Tree - Fenwick Tree """

    def __init__(self, arr):

        self.n = len(arr)
        self.bit_array = [0]+arr

        for i in range(1,self.n+1):
            idx = i + ( i & (-i) )
            while idx<=self.n:
                self.bit_array[idx] += arr[i-1]
                idx = idx + ( idx & (-idx))

        print('Array: [{}]'.format(', '.join(map(str, self.bit_array))))

    def prefix_query(self, n):
        idx  = n+1
        sum =0
        while idx>0:
            sum += self.bit_array[idx]
            idx -= (idx & (-idx))
        return sum

    def range_query(self, from_idx, to_idx):
        if from_idx ==0:
            return self.prefix_query(to_idx)
        else:
            return self.prefix_query(to_idx) - self.prefix_query(from_idx)

    def update(self, idx, num):

        idx += 1
        while idx <=self.n:
            self.bit_array[idx] += num
            idx += (idx & (-idx))


if __name__== "__main__":

    arr = [1, 7, 3, 0, 5, 8, 3, 2, 6, 2, 1, 1, 4, 5]
    fenwick = BIT(arr)


    print('Prefix sum of first {} elements: {}'.format(13, fenwick.prefix_query(12)))
    print('Prefix sum of first {} elements: {}'.format(7, fenwick.prefix_query(6)))
    print('Range sum from pos {} to pos {}: {}'.format(1, 5, fenwick.range_query(1, 5)))

    print('Add {} to element at idx {}\n \n '.format(2, 4))
    fenwick.update(4, 2)

    print('Prefix sum of first {} elements: {}'.format(13, fenwick.prefix_query(12)))
    print('Prefix sum of first {} elements: {}'.format(7, fenwick.prefix_query(6)))
    print('Range sum from pos {} to pos {}: {}'.format(1, 5, fenwick.range_query(1, 5)))

File: cpp/practise/binary-indexed-tree.cpp
#include <iostream>
#include <vector>
#include <iomanip>

using namespace std;

class BIT{

    public:

        BIT(vector<int> arr, int n){

            bit_array = vector<int>(n+1, 0);

            for(int i=1; i<=n; i++){

                int idx = i;
                while(idx<=n){
                    bit_array[idx] +=arr[i-1];
                    idx += idx &(-idx);
                }
            }

            for(int i=0; i<=n; i++){
                cout<< std::setw(2)<< bit_array[i]<<" ";
            }

            cout<<endl;

        }

        int prefix_query(int idx){
            idx +=1;
            int sum =0;

            while(idx> 0){
                sum += bit_array[idx];
                cout<<sum<<"==>"<<idx <<endl;
                idx -= idx & (-idx);
            }

            return sum;
        } //prefix_query

        int range_query(int from_idx, int to_idx){

            if(from_idx ==0)
                return prefix_query(to_idx);
            else
                return (prefix_query(to_idx) -  prefix_query(from_idx));
        } //range_query

        //int update(int idx, int value)


    private:

        vector<int> bit_array;

};

void print(BIT &bit, int n) {


    //std::cout << "Prefix sum of first 13 elements: \t" << bit.range_query(0,12) << std::endl;
	//std::cout << "Prefix sum of first 7 elements: \t" << bit.prefix_query(6) << std::endl;
	std::cout << "Range sum from index 1 to index 5: \t" << bit.range_query(1, 5) << std::endl;
    std::cout << std::endl;
}

int main(){

    vector<int> arr{ 1, 7, 3, 0, 5, 8, 3, 2, 6, 2, 1, 1, 4, 5 };
    int n = arr.size();

    //BIT bit = new BIT(array, n);

    BIT bit(arr, n);

    print(bit, n);



    return 0;
}

